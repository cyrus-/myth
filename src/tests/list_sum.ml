let s = "
type nat =
  | O
  | S of nat

type list =
  | Nil
  | Cons of nat * list

let rec fold (l:list) (f:nat -> nat -> nat) (acc:nat) : nat =
  match l with
  | Nil -> acc
  | Cons (x, l) -> fold l f (f acc x)
;;

let rec add (n1:nat) (n2:nat) : nat =
  match n1 with
  | O -> n2
  | S (n1) -> S (add n1 n2)
;;

let list_sum : list -> nat |>
  { [] => 0
  | [1] => 1
  | [2; 1] => 3
  } = ?"
