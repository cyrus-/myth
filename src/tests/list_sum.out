(* Generated Thu Nov 13 00:35:52 2014 *)
let list_sum : list -> nat =
  fun (l1:list) -> fold l1 add O
;;
